File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);

            if(c=='(' || c=='[' || c=='{'){
                stack.push(c);
            } else {
                if (stack.isEmpty()) {return false;}
                char open = stack.pop();
                if ( (c==')' && open!='(') 
                || (c==']' && open!='[')
                || (c=='}' && open!='{') ){
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for c in s:
            if (c=='(') or (c=='[') or (c=='{'):
                stack.append(c)
            else:
                if not stack:
                    return False
                open_char = stack.pop()
                if (c==")" and open_char!="(") or (c=="]" and open_char!="[") or (c=="}" and open_char!="{"):
                    return False
        
        return stack == []